keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -543,56 +686,132 @@ type ReplicaLimit struct {
 }
 ```
 
-For comparison (provided by @liggitt):
+`MatchPolicy` will work the same as for admission webhooks. It will default to
+`Equivalent` but may be set to `Exact`. See "Use Case: Multiple policy
+definitions for different versions of CRD" for an explanation of why we need
+`MatchPolicy`.
 
-- RBAC, priority & fairness, and admission webhook configurations all use
-ordered lists of rules, and first match allows/matches.
-- Audit configuration also uses an ordered list of rules, but can make a match
-exclude auditing... this approach could address the "ability to skip specific
-resource types" mentioned above.
+xref:
+https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules
 
-TODO: What exact matching mechanisms will be supported? Right now this KEP
-shows the use of the same match rules that admission webhooks use.
 
-TODO: Should policy definitions also have match criteria? Match criteria at the
-policy definition level could be used constrain what resources the policy may
-apply to across the entire cluster (like the above "kind constraint" but more
-generally). This would allow policy authors to constrain use of the policy to
-the resources it was designed for.
+Alternative considered: Define match criteria with CEL expressions
 
-TODO: Should "standalone policies" be supported? I.e. should it be possible to
-declare a policy definition that has no `spec.config` but has it's own match
-criteria? Right now a policy requires a minimum of three resources to do
-anything (policy definition, configuration CRD, at least one policy
-configuration).
+Pros:
 
-`MatchPolicy` will implicitly be `Equivalent`, and will not be configurable,
-but it will be possible to opt-out of the behavior by inspecting the
-GVK of the admitted resource in the CEL expression.
+- CEL expressions can be used to declare far more sophisticated match criteria.
 
-xref:
-https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules
+Cons:
+
+- CEL expressions can evalaute to an error while API types only evaluate to
+  "matches" / "does not match" result.
+- We intend to allow policy configuration resources declare match criteria,
+  making it difficult to pre-compile or type-check CEL expressions.
+
+##### Exclude matching
+
+Benefits:
+
+- Exclude rules make it easier for policy definition authors to do things like
+  exclude kube-system or other special purpose namespaces from policies that apply
+  to all "normal" namespaces.
+- Default rules make it easier to apply principal of least privilege to
+  policies. I.e. ensure that newly created resources default to a least
+  privilege state until configured otherwise.
+
+Exclude rules can be modeled after the approach exstablished in Audit where
```

Comment:
I think there are at least two approaches to exclude: interactive and non-interactive. This interactive one is more problematic, and I don't see the non-interactive one considered.

This approach appears to describe a resource that can preempt other resources based off priority. This complicates the ability to reason about policy, since adding a new policy may not actually be effective due to priority, and also makes the eventual consistency story harder to reason about: in a world where policies don't interact then a non-ingested policy means your cluster is strictly less locked-down than specified, that same assumption can't be made if inter-policy interactions are added.

Another approach is to add exclude rules that apply only to the policy in-question e.g. `match.excludedNamespaces`, but don't affect other policies. This is essentially a denormalized version of what's currently suggested.

It may make sense to have global exclusion objects, per https://github.com/kubernetes/kubernetes/issues/92157 but that would be to remove control of certain resources from the user space entirely, so cluster admins can ensure their "superroot" operations can't be interfered with.

Reply:
I'm not familiar with the notion of interactive and non-interactive.

Unless I'm misunderstanding how Audit excludes work, the `level=None` has a priority (based on order in the match list), but only impacts the list of match rules contained in that Audit resource, not across resources.  I agree that any exclude rules that applies across priority resources is very problematic.

My mental model for what I want matching criteria to do is basically a query filter (e.g. SQL WHERE clause).  `(<some criteria>) AND (<some criteria>) AND NOT (<some exclude criteria>)`.

To use Audit match rules I somehow have to express this in a list of match criteria that short-circuits on the first match. For cases with very few exclude rules this seems manageable?  I'm open to ideas here. I'll admit I'm biasing toward using approaches already in use in the Kubernetes API, but I think if we find them deficient, we should do something better.

Reply:
https://github.com/kubernetes/enhancements/pull/3492#discussion_r970211846